src/webgpu/shader/types.ts

```diff
@@ -261,51 +299,64 @@ export function* generateTypes({
 
   // Array types
   if (containerType === 'array') {
-    // Buffer affective binding size must be a multiple of 4. Adjust array length as needed.
-    let arrayLength = kDefaultArrayLength;
-    if (
-      addressSpace === 'storage' &&
-      scalarInfo.layout !== undefined &&
-      scalarInfo.layout.alignment % 4 !== 0
-    ) {
-      arrayLength = align(arrayLength, 4);
+    let arrayElemType: string = scalarType;
+    let arrayElementCount: number = kDefaultArrayLength;
+    let supportsAtomics = scalarInfo.supportsAtomics;
+    let layout: undefined | AlignmentAndSize = undefined;
+    if (scalarInfo.layout) {
+      // Compute the layout of the array type.
+      // Adjust the array element count or element type as needed.
+      if (addressSpace === 'uniform') {
+        // Use a vec4 of the scalar type, to achieve a 16 byte alignment without internal padding.
+        // This works for 4-byte scalar types, and does not work for f16.
+        // It is the caller's responsibility to filter out the f16 case.
+        assert(!isAtomic, 'the uniform case is making vec4 of scalar, which cannot handle atomics');
+        arrayElemType = `vec4<${baseType}>`;
+        supportsAtomics = false;
+        const arrayElemLayout = vectorLayout('vec4', baseType) as AlignmentAndSize;
```

Comment:
nit: `!` is safer than `as`
```suggestion
        const arrayElemLayout = vectorLayout('vec4', baseType)!;
```